cmd/geoipupdate/main.go

```diff
@@ -0,0 +1,428 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	flock "github.com/theckman/go-flock"
+)
+
+// Version is the program's version number.
+const Version = "0.0.1"
+
+func main() {
+	log.SetFlags(0)
+
+	args := getArgs()
+
+	config, err := NewConfig(args.ConfigFile, args.DatabaseDirectory)
+	if err != nil {
+		log.Fatalf("Error loading configuration file: %+v", err)
+	}
+
+	lock, err := setup(config, args.Verbose)
+	if err != nil {
+		log.Fatalf("Error preparing to update: %+v", err)
+	}
+	defer func() {
+		if err := lock.Unlock(); err != nil {
+			log.Fatalf("Error unlocking lock file: %+v", errors.Wrap(err, "unlocking"))
+		}
+	}()
+
+	if err := run(config, args.Verbose); err != nil {
+		log.Fatalf("Error retrieving updates: %+v", err)
+	}
+}
+
+func setup(
+	config *Config,
+	verbose bool,
+) (*flock.Flock, error) {
+	if err := maybeSetProxy(config); err != nil {
+		return nil, err
+	}
+
+	if err := checkEnvironment(config); err != nil {
+		return nil, err
+	}
+
+	lock := flock.NewFlock(config.LockFile)
+	ok, err := lock.TryLock()
+	if err != nil {
+		return nil, errors.Wrap(err, "error acquiring a lock")
+	}
+	if !ok {
+		return nil, errors.Errorf("could not acquire lock on %s", config.LockFile)
+	}
+	if verbose {
+		log.Printf("Acquired lock file lock (%s)", config.LockFile)
+	}
+
+	return lock, nil
+}
+
+var client = &http.Client{
+	Timeout: 30 * time.Minute, // Allow for slow downloads
+}
+
+func maybeSetProxy(config *Config) error {
+	if config.Proxy == "" {
+		return nil
+	}
+	proxy := config.Proxy
+
+	if strings.Index(proxy, ":") == -1 {
```

Comment:
Was it possible to pass in the auth using `Proxy` previous, e.g., `user:pass@127.0.0.1`?

Reply:
Unfortunately it appears so, despite the relevant cURL docs saying nothing about that (https://curl.haxx.se/libcurl/c/CURLOPT_PROXY.html). I'll see about supporting it too.